### ISSUE TITLE
Too many 502 responses lately

### ISSUE THREAD

**Issue by Daniel-Mietchen** [arthurpsmith/author-disambiguator#163]
I am getting HTTP 502 responses most of the time when trying to access Author Disambiguator, but especially if I am trying to access it two or more times in parallel.

Has anything changed in the setup?

**Comment by wafflefin** [comment]
I have been experiencing this extensively on two different connections on various devices for the past week or so. Even reducing the limit on most pages is not sufficient to mitigate it. Even with a limit of 50 or 10, pages that complete take an extraordinary long time.

**Comment by axel-wilke** [comment]
Same for me.

**Comment by arthurpsmith** [comment]
Hi all - sorry about this, I'd noticed it getting worse recently myself. There are (I believe) 4 servers available to respond to requests, but since the requests do a series of what can be complex WDQS queries, the response can be slow. I believe what's happening is a series of super-slow requests are clogging up the available servers, and so the next requests have to wait, and may time out generating these 502s.

Accessing it 2 or more times in parallel does add to the load causing trouble. If you are aware that some of your requests are always slow (for a particular sort of name for example) it would probably be helpful to postpone those for a later time whenever you hit a 502.

The simplest solution is to increase capacity on toolforge so more servers are responding to requests. I'll look into this.

The best long-term solution is to figure out some way to eliminate the slow requests. One route to do that is to move most of the query logic into client-side javascript so it's running on the client browser; the response from this server would be quick and your browser would be directly contacting WDQS for the slow bits.
Another option would be to respond somehow asynchronously - that is have a quick response that just polls waiting for the actual results to be available. I'm already doing a quick response/poll for running the update batches (after a previous episode with 502s made it clear we needed to do that) so I know it's at least possible to run things that way.

Anyway, doing the long-term things will definitely taking quite a bit of coding one way or another, so I'll see what we can do about the pool of servers first.

**Comment by arthurpsmith** [comment]
Hmm,  it's possible there's only one server  responding. I definitely need to address this.  In principle it's simple, however, with OAuth I need stateful sessions which get complicated  across  multiple servers (I attempted to add additional  kubernetes  "pods" this  morning and got  OAuth  errors!).  I  have to think and research on this a  bit.

**Comment by arthurpsmith** [comment]
Ok, it's  running now with  4 kubernetes "pods" -  this should  greatly reduce the incidence of this problem. There may  still be some issues to sort out with batch handling though.
